cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -426,11 +439,69 @@ public Set<ProgramId> filterProgramsExistence(Collection<ProgramId> programIds)
   }
 
 
-  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec)
+  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec,
+                                long creationTimeMillis, @Nullable String author, @Nullable String changeSummary)
     throws IOException {
-    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec)));
+    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec,
+                                                                                              changeSummary,
+                                                                                              creationTimeMillis,
+                                                                                              author))
+      , creationTimeMillis, author, changeSummary);
   }
 
+  public void createApplicationVersion(ApplicationId id, ApplicationMeta meta, @Nullable String parentVersion)
+    throws IOException, BadRequestException {
+    // Fetch the latest version
+    ApplicationMeta latest = getLatest(id.getNamespaceId(), id.getApplication()).stream().findAny().orElse(null);
+    //  If latest is null and app exists (legacy versions) then fetch -SNAPSHOT app
+    if (latest == null) {
+      latest = getApplication(id.getNamespace(), id.getApplication(), ApplicationId.DEFAULT_VERSION);
+    }
+    // If -SNAPSHOT version does not exist, return the app corresponding to the smallest version-id string
```

Comment:
don't we want the 'highest' versionid?

Also, can we skip the '-snapshot' lookup and just do this scan?

Reply:
agree, we perhaps can use the scan with sorted creation DESC time method

Reply:
so the pick is based on `version-id` (not creation time) - this is the case when existing apps - don't have the latest column and we need to make a selection (pick a row) to update the existing latest field to false. I don't think it should matter if we pick the smallest or the highest version id